python/tvm/autotvm/task/task.py

```diff
@@ -0,0 +1,340 @@
+        if not check_correctness:
+            return
+        with _target.create("llvm"):
+            s, args = self.instantiate(self.config_space.get(0))
+        func = build(s, args, "llvm")
+        tvm_buf = [nd.array(x) for x in self.ref_input]
+        func(*tvm_buf)
+        self.ref_output = [x.asnumpy() for x in tvm_buf]
+
+    def init_space(self, target, target_host=None, template_key=None):
+        """Initialize task for a specific target.
+        This function should be called before a task is used by tuner.
+
+        Parameters
+        ----------
+        target: tvm.target.Target
+            The hardware target
+        target_host: tvm.target.Target, optional
+            The host side target
+        template_key: str, optional
+            The template key to use
+        """
+        if isinstance(target, str):
+            target = _target.create(target)
+
+        self.config_space = ConfigSpace()
+        self.config_space.template_key = template_key or ""
+
+        ctx = ApplyConfig(self.config_space)
+        with ctx:
+            with target:
+                sch, _ = self.func(*self.args, **self.kwargs)
+                self.config_space.code_hash = getattr(sch, 'code_hash', None)
+        self.workload = ctx.last_workload
+        self.flop = self.config_space.flop or compute_flop(sch)
+
+        self.target = target
+        self.target_host = target_host
+
+    def instantiate(self, config):
+        """Instantiate a task function (template) with a config.
+
+        Parameters
+        ----------
+        config: template.ConfigEntity
+            parameter config for this template
+
+        Returns
+        -------
+        sch: tvm.schedule.Schedule
+            The tvm schedule
+        arg_bufs: Array of tvm.tensor.Tensor
+            The input/output buffers
+        """
+        config.flop = 0
+        with ApplyConfig(config):
+            sch, arg_bufs = self.func(*self.args, **self.kwargs)
+        if not self.flop:
+            self.flop = config.flop or compute_flop(sch)
+        return sch, arg_bufs
+
+    def __repr__(self):
+        return "Task(func_name=%s, args=%s, kwargs=%s, workload=%s)" % (
+            self.name, self.args, self.kwargs, self.workload
+        )
+
+
+TASK_TABLE = {
+}
+
+def register(name, func=None, override=False):
+    """Register a task function.
+
+    Parameters
+    ----------
+    name : str
+        The name to identify the task.
+    func : callable
+        The function to be registered.
+    override : bool
+        Whether override existing registration.
+
+    Returns
+    -------
+    func: callable
+        The registered function
+    """
+    def _do_reg(myf):
+        if name in TASK_TABLE and not override:
+            raise ValueError(
+                "Key %s is already registered" % name)
+        TASK_TABLE[name] = myf
+        return myf
+    if func:
+        return _do_reg(func)
+    return _do_reg
+
+
+def create(func_name, args):
```

Comment:
create and create_task can be super confusing, I would prefer to rename create_task-> create and remove create. 

Allow pass in None in target, and call init_space later which returns a new task